tests/clickhouse/optimize/test_optimize.py

```diff
@@ -244,6 +246,100 @@ def test_metrics_tags(
         assert _get_metrics_tags(table, host) == expected
 
 
+test_data = [
+    pytest.param(
+        StorageKey.ERRORS,
+        lambda dt: InsertBatch(
+            [
+                {
+                    "event_id": str(uuid.uuid4()),
+                    "project_id": 1,
+                    "group_id": 1,
+                    "deleted": 0,
+                    "timestamp": dt,
+                    "retention_days": settings.DEFAULT_RETENTION_DAYS,
+                }
+            ],
+            None,
+        ),
+        last_midnight
+        + timedelta(hours=settings.PARALLEL_OPTIMIZE_JOB_START_TIME)
+        + timedelta(minutes=30),
+        id="errors parallel",
+    )
+]
+
+
+class TestOptimizeFrequency:
+    @pytest.mark.clickhouse_db
+    @pytest.mark.redis_db
+    @pytest.mark.parametrize(
+        "storage_key, create_event_row_for_date, current_time",
+        test_data,
+    )
+    @patch("snuba.clickhouse.optimize.optimize.optimize_partitions")
+    def test_optimize_partitions_is_called_the_same_number_of_times_as_number_of_partitions(
+        self,
+        mock_optimize_partitions: Mock,
+        storage_key: StorageKey,
+        create_event_row_for_date: Callable[[datetime], InsertBatch],
+        current_time: datetime,
+    ) -> None:
+        mock_optimize_partitions.side_effect = optimize_partitions
+
+        storage = get_writable_storage(storage_key)
+        cluster = storage.get_cluster()
+        clickhouse = cluster.get_query_connection(ClickhouseClientSettings.OPTIMIZE)
+        table = storage.get_table_writer().get_schema().get_local_table_name()
+        database = cluster.get_database()
+
+        base = datetime(1999, 12, 26)  # a sunday
+        base_monday = base - timedelta(days=base.weekday())
+        write_processed_messages(storage, [create_event_row_for_date(base)])
+        write_processed_messages(storage, [create_event_row_for_date(base)])
+
+        for i in range(1, 10):
+            earlier = base_monday - timedelta(days=i * 31)
+            earlier_monday = earlier - timedelta(days=earlier.weekday())
+            write_processed_messages(
+                storage, [create_event_row_for_date(earlier_monday)]
+            )
+            write_processed_messages(
+                storage, [create_event_row_for_date(earlier_monday)]
+            )
+
+        scheduler = OptimizeScheduler(2)
+        tracker = OptimizedPartitionTracker(
+            redis_client=redis_client,
+            host="some-hostname.domain.com",
+            port=9000,
+            database=database,
+            table=table,
+            expire_time=datetime.now() + timedelta(minutes=10),
+        )
+        partitions = optimize.get_partitions_to_optimize(
+            clickhouse, storage, database, table
+        )
+
+        tracker.update_all_partitions([part.name for part in partitions])
```

Comment:
If I comment out this line, the test still passes. Do we need it? I see that this line causes some state to be written to redis, but I don't understand how that state fits in with the logic we changed. Could you explain that?